### DISCUSSION TITLE
Is there a way to get the pathname with a locale out of url?

### DISCUSSION THREAD

**Discussion by GabenGar** [vercel/next.js#36506] [Help]
`Link` seems to do it automatically, but how do I do it outside of React rendering tree?

**Comment by icyJoseph** [comment]
You can use the `useRouter().locale` option, or import the `Router` instance, and read the locale from there. Make sure you only use these on the client.

I am not sure if you want to get the pathname with the locale, or just the locale and just the pathname?

You can also access the locale in _middleware with `req.nextUrl.locale`.

Where and how exactly do you want to use it?

**Reply by GabenGar** [reply]
> I am not sure if you want to get the pathname with the locale, or just the locale and just the pathname?

All of the above I guess, depends on the context. But more importantly I'd like to create locale-aware links with a single function, preferably a semi-isomorphic one between server and client.

>Where and how exactly do you want to use it?

- canonical links for SEO tags, I was feeding `useRouter().pathname` to them and the result is not locale-aware.
- `Redirect` results from `GetServerSideProps()`, these aren't locale-aware either.
- also I'd like to get a locale info for various modules, such as date formatting, both on server and client.

**Comment by icyJoseph** [comment]
>canonical links for SEO tags, I was feeding useRouter().pathname to them and the result is not locale-aware.

I think `useRouter().locale` should work for these, right? (added to `pathname`), you could also use `asPath`, which, except for the default locale, should contain the locale.

>Redirect results from GetServerSideProps(), these aren't locale-aware either.

Context contains the locale in getServerSideProps. You'll need to prepend that to the redirect url though.

>also I'd like to get a locale info for various modules, such as date formatting, both on server and client.

I think the Router instance could work here, as long as these modules are strictly used client side. I haven't done this myself, but I reckon, at the moment you call a function from this module and read `Router.locale` you should get a proper value.

The bad option would be to always take the locale as a parameter.

**Reply by GabenGar** [reply]
> I think `useRouter().locale` should work for these, right?

It should, but then I have to write a separate function to create a URL out of that. Said function will also have to accept `defaultLocale` to figure if it has to prepend anything at all. I will write it eventually, but it sounds like duplicate work seeing as NextJS knows better about its own URLs, their locales and how to deal with them. 

>(added to `pathname`), you could also use `asPath`, which, except for the default locale, should contain the locale.

Does it? The [docs say a different thing](https://nextjs.org/docs/api-reference/next/router#router-object), or at least a bit too vague on this detail:

>`asPath`: `String` - The path (including the query) shown in the browser without the configured `basePath` or `locale`.

Also "including the query" is a deal breaker for canonical links, since they are constructed without queries. So that's additional processing for removing them from this string.

> Context contains the locale in getServerSideProps. You'll need to prepend that to the redirect url though.

That's a separate function which, much like the the canonical one, has to take `defaultLocale` into account. And unlike SEO tags it's a part of NextJS and breaks the interaction with middlewares, which do respect locale without explicitly passing it in redirects. Should I create a separate issue for that?

>The bad option would be to always take the locale as a parameter.

It's not bad, but this argument should be optional.

**Comment by GabenGar** [comment]
The issue with the default import from `next/router` is it requires to be VERY careful how it is called. I thought `useRouter()` was also firing on server but apparently it's not so if I want to statically create a link in the page component I have to pass at least `locale` and `defaultLocale` to the page props too (or create a link within `GetServerSideProps()`/`GetStaticProps()`), which definitely doesn't reduce verbosity.
